File: fboss/platform/bsp_tests/utils/kmod_utils.py
# pyre-strict
import json

from fboss.platform.bsp_tests.utils.cmd_utils import check_cmd, run_cmd
from fboss.platform.platform_manager.platform_manager_config.types import BspKmodsFile
from thrift.py3 import deserialize, Protocol


def build_util_path(vendor_name: str) -> str:
    kern_ver = run_cmd(["uname", "-r"]).stdout.decode().strip()
    return f"/usr/local/{vendor_name}_bsp/{kern_ver}"


def read_kmods(vendor_name: str) -> BspKmodsFile:
    kmod_path = f"{build_util_path(vendor_name)}/kmods.json"

    with open(kmod_path, "r") as f:
        json_data = json.load(f)

    return deserialize(
        BspKmodsFile, json.dumps(json_data).encode("utf-8"), protocol=Protocol.JSON
    )


def load_kmods(kmods: BspKmodsFile) -> None:
    for kmod in kmods.sharedKmods:
        check_cmd(["modprobe", kmod])
    for kmod in kmods.bspKmods:
        check_cmd(["modprobe", kmod])


def unload_kmods(kmods: BspKmodsFile) -> None:
    for kmod in kmods.bspKmods:
        check_cmd(["modprobe", "-r", kmod])
    for kmod in kmods.sharedKmods:
        check_cmd(["modprobe", "-r", kmod])


def fbsp_remove(vendor_name: str) -> None:
    check_cmd([f"{build_util_path(vendor_name)}/fbsp-remove.sh", "-f"])


def get_loaded_kmods(expected_kmods: BspKmodsFile) -> list[str]:
    result = run_cmd("lsmod").stdout.decode()
    expected_kmod_list = set(
        list(expected_kmods.bspKmods) + list(expected_kmods.sharedKmods)
    )
    kmods = []
    for line in result.split("\n"):
        parts = line.strip().split()
        if len(parts) == 0:
            continue
        kmod = parts[0]
        if kmod in expected_kmod_list:
            kmods.append(kmod)
    return kmods

File: fboss/platform/bsp_tests/utils/KmodUtils.cpp
// Copyright (c) Meta Platforms, Inc. and affiliates.

#include "fboss/platform/bsp_tests/utils/KmodUtils.h"

#include <fmt/format.h>
#include <folly/Format.h>
#include <folly/logging/xlog.h>
#include <gtest/gtest.h>
#include <re2/re2.h>
#include <set>
#include <sstream>
#include <string>
#include <vector>

#include "fboss/platform/helpers/PlatformUtils.h"
#include "fboss/platform/platform_manager/PkgManager.h"

namespace facebook::fboss::platform::bsp_tests {

void KmodUtils::loadKmods(const BspKmodsFile& kmods) {
  // Load shared kmods first
  for (const auto& kmod : *kmods.sharedKmods()) {
    auto result = PlatformUtils().execCommand(fmt::format("modprobe {}", kmod));
    EXPECT_EQ(result.first, 0) << "Failed to load shared kmod: " << kmod;
  }

  // Then load BSP kmods
  for (const auto& kmod : *kmods.bspKmods()) {
    auto result = PlatformUtils().execCommand(fmt::format("modprobe {}", kmod));
    EXPECT_EQ(result.first, 0) << "Failed to load BSP kmod: " << kmod;
  }
}

void KmodUtils::unloadKmods(const BspKmodsFile& kmods) {
  // Unload non-shared kmods first
  for (const auto& kmod : *kmods.bspKmods()) {
    auto result =
        PlatformUtils().execCommand(fmt::format("modprobe -r {}", kmod));
    EXPECT_EQ(result.first, 0) << "Failed to unload BSP kmod: " << kmod;
  }

  // Then unload shared kmods
  for (const auto& kmod : *kmods.sharedKmods()) {
    auto result =
        PlatformUtils().execCommand(fmt::format("modprobe -r {}", kmod));
    EXPECT_EQ(result.first, 0) << "Failed to unload shared kmod: " << kmod;
  }
}

// Run fbsp-remove.sh script
void KmodUtils::fbspRemove(
    const platform_manager::PlatformConfig& platformConfig) {
  // Find fbsp-remove.sh script path
  constexpr auto kremoveScriptPath = "/usr/local/{}_bsp/{}/fbsp-remove.sh";
  const re2::RE2 kBspRpmNameRe = "(?P<KEYWORD>[a-z]+)_bsp_kmods";
  std::string keyword{};
  re2::RE2::FullMatch(
      *platformConfig.bspKmodsRpmName(), kBspRpmNameRe, &keyword);
  std::string fbspPath = fmt::format(
      kremoveScriptPath,
      keyword,
      platform_manager::package_manager::SystemInterface()
          .getHostKernelVersion());

  XLOG(DBG) << "Running fbsp-remove: " << fbspPath;
  auto result = PlatformUtils().execCommand(fmt::format("{} -f", fbspPath));
  EXPECT_EQ(result.first, 0) << "Failed to run fbsp-remove.sh";
}

// Get list of loaded kernel modules from `lsmod` that match expected modules
std::vector<std::string> KmodUtils::getLoadedKmods(
    const BspKmodsFile& expectedKmods) {
  auto lsmodResult = PlatformUtils().execCommand("lsmod");
  EXPECT_EQ(lsmodResult.first, 0) << "Failed to run lsmod";
  const auto& lsmodOutput = lsmodResult.second;

  std::set<std::string> expectedKmodSet;
  for (const auto& kmod : *expectedKmods.bspKmods()) {
    expectedKmodSet.insert(kmod);
  }
  for (const auto& kmod : *expectedKmods.sharedKmods()) {
    expectedKmodSet.insert(kmod);
  }

  // Parse lsmod output and find matching kmods
  std::vector<std::string> loadedKmods;
  std::istringstream iss(lsmodOutput);
  std::string line;

  // Skip header line
  std::getline(iss, line);

  while (std::getline(iss, line)) {
    std::istringstream lineStream(line);
    std::string kmod;
    lineStream >> kmod;

    if (!kmod.empty() && expectedKmodSet.find(kmod) != expectedKmodSet.end()) {
      loadedKmods.push_back(kmod);
    }
  }

  return loadedKmods;
}

} // namespace facebook::fboss::platform::bsp_tests
